modules/i2s/src/i2s_tdm_slave.c

```diff
@@ -0,0 +1,222 @@
+    port_clear_buffer(ctx->p_fsync);
+    
+    if (ctx->slave_bclk_polarity == I2S_SLAVE_SAMPLE_ON_BCLK_FALLING) {
+        port_set_invert(ctx->p_bclk);
+    } else {
+        port_set_no_invert(ctx->p_bclk);
+    }
+    clock_start(ctx->bclk);
+}
+
+static void i2s_tdm_slave_deinit_resources(
+        i2s_tdm_ctx_t *ctx)
+{
+    clock_disable(ctx->bclk);
+    port_disable(ctx->p_bclk);
+
+    for(int i=0; i<ctx->num_out; i++) {
+        port_disable(ctx->p_dout[i]);
+    }
+    for(int i=0; i<ctx->num_in; i++) {
+        port_disable(ctx->p_din[i]);
+    }
+}
+
+void i2s_tdm_slave_tx_16_thread(
+        i2s_tdm_ctx_t *ctx)
+{
+    uint32_t out_samps[I2S_TDM_MAX_CH_PER_FRAME];
+    uint32_t fsync_val = 0;
+
+    while(1) {
+        xassert(ctx->num_out == 1);
+
+        if (ctx->i2s_cbg->init != NULL) {
+            ctx->i2s_cbg->init(ctx->i2s_cbg->app_data, NULL);
+        }
+
+        i2s_tdm_slave_init_resources(ctx);
+
+        /* Note we init the resources first to allow port delays etc. to be modified
+           after ports and clocks have been initialised */
+        if (ctx->tdm_post_port_init != NULL) {
+            ctx->tdm_post_port_init(ctx);
+        }
+
+        /* Get first frame data */
+        ctx->i2s_cbg->send(ctx->i2s_cbg->app_data, ctx->ch_per_frame, (int32_t*)out_samps);
+
+        uint32_t port_frame_time = (ctx->ch_per_frame * ctx->word_len);
+
+        /* Wait for first fsync rising edge to occur */
+        ctx->fysnch_error = false;
+        port_set_trigger_in_equal(ctx->p_fsync, 0);
+        (void) port_in(ctx->p_fsync);
+        port_set_trigger_in_equal(ctx->p_fsync, 1);
+        (void) port_in(ctx->p_fsync);
+        port_timestamp_t fsync_edge_time = port_get_trigger_time(ctx->p_fsync);
+        port_clear_trigger_in(ctx->p_fsync);
+
+        /* Setup trigger times */
+        port_set_trigger_time(ctx->p_fsync, port_frame_time + fsync_edge_time);
+        port_set_trigger_time(ctx->p_dout[0], port_frame_time + fsync_edge_time + ctx->tx_offset);
+
+        port_out(ctx->p_dout[0], bitrev(out_samps[0]));
+        fsync_val = port_in(ctx->p_fsync);
+        fsync_edge_time = port_get_trigger_time(ctx->p_fsync);
+        port_set_trigger_time(ctx->p_fsync, port_frame_time + fsync_edge_time);
+        
+        for (int i=1; i<ctx->ch_per_frame; i++) {
+            port_out(ctx->p_dout[0], bitrev(out_samps[i]));
+        }
+
+        while(1) {
+            /* Get frame data and tx */
+            ctx->i2s_cbg->send(ctx->i2s_cbg->app_data, ctx->ch_per_frame, (int32_t*)out_samps);
+
+            port_out(ctx->p_dout[0], bitrev(out_samps[0]));
+            fsync_val = port_in(ctx->p_fsync);
+            fsync_edge_time = port_get_trigger_time(ctx->p_fsync);
+            port_set_trigger_time(ctx->p_fsync, fsync_edge_time + port_frame_time);
+
+            /* Note: Still possible for us to alias, but this will catch nonperiod drifting */
+            if (fsync_val != 1) {
+                ctx->fysnch_error = true;
```

Comment:
From what I can see, no method exists to report a synchronisation error or to clear this flag for one.

Reply:
the fysnch_error is a sticky flag which gets cleared when TDM restarts (and resynchs). It is up to the application writer to monitor this if needed. This is an enhancement I added to TDM slave over the I2S API which I found useful during development. Otherwise, you just get the call to the init() callback with no way of know why you are there.

I have added some notes in the API to help with this.